### PR TITLE
feat: add `noirc_abi_wasm` crate for ABI encoding in JS

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves 

## Summary\*

Draft PR for tracking development of the Noir ABI encoding wasm package.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_abi_wasm/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+```bash
+nix build -L github:noir-lang/noir/master#abi_wasm
+```
+
+If you are within the noir repo and would like to build local changes, you can use:
+
+```bash
+nix build -L #abi_wasm
+```
```
Comment (TomAFrench) [inline crates/noirc_abi_wasm/README.md:17]: TODO: nix config

File: crates/noirc_abi_wasm/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+use gloo_utils::format::JsValueSerdeExt;
+use wasm_bindgen::{prelude::wasm_bindgen, JsValue};
+
+mod js_witness_map;
+mod temp;
+
+use js_witness_map::JsWitnessMap;
+use temp::{input_value_from_json_type, JsonTypes};
+
+#[wasm_bindgen(js_name = abiEncode)]
```
Comment (TomAFrench) [inline crates/noirc_abi_wasm/src/lib.rs:20]: TODO: document these.

**Comment by TomAFrench** [comment]
@kobyhallx I forgot to mention, I'm using this PR for developing the abi encoding package.

**Comment by TomAFrench** [comment]
Looks like `result` is being created up in the workspace root rather than in the `noirc_abi_wasm` crate.

**Comment by kevaundray** [comment]
@TomAFrench what is the status of this and does it have a corresponding issue?

**Comment by TomAFrench** [comment]
@kevaundray I've added an issue for this. This PR is ongoing as time allows.

File: flake.nix [hunk 3]
```diff
@@ -274,7 +288,10 @@
           toml2json
           llvmPackages.lldb # This ensures the right lldb is in the environment for running rust-lldb
           wasm-bindgen-cli
+          binaryen
           jq
+          rust-bin.stable."1.66.1".default
+          rust-analyzer
```
Comment (TomAFrench) [inline flake.nix:294]: @kobyhallx Why was this added to this PR? Not sure why we'd need to modify the CLI nix config as this PR only addresses the new ABI package.

File: .devcontainer.json
Comment (TomAFrench) [inline .devcontainer.json:1]: Can you motivate this @kobyhallx? If so we should split this out so it's not blocked by the ABI crate.

Comment (TomAFrench) [reply .devcontainer.json:1]: I've removed this.

File: crates/noirc_abi_wasm/test/browser/abi_encode.test.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { expect } from "@esm-bundle/chai";
+import initACVM, { abiEncode, abiDecode, WitnessMap } from "../../../../result";
+import { DecodedInputs } from "../types";
+
+beforeEach(async () => {
+  await initACVM();
```
Comment (TomAFrench) [inline crates/noirc_abi_wasm/test/browser/abi_encode.test.ts:6]:
```suggestion
import initNoirAbi, { abiEncode, abiDecode, WitnessMap } from "../../../../result";
import { DecodedInputs } from "../types";

beforeEach(async () => {
  await initNoirAbi();
```

File: flake.nix [hunk 5]
```diff
@@ -167,8 +167,8 @@
         ];
 
         buildInputs = [
-          pkgs.llvmPackages.openmp
-          pkgs.barretenberg
+          # pkgs.llvmPackages.openmp
+          # pkgs.barretenberg
```
Comment (TomAFrench) [inline flake.nix:171]: commented code

File: flake.nix [hunk 6]
```diff
@@ -325,6 +340,13 @@
           llvmPackages.lldb # This ensures the right lldb is in the environment for running rust-lldb
           wasm-bindgen-cli
           jq
+          binaryen
+          yarn
+          rust-bin.stable."1.66.1".default
+          rust-analyzer
+          rustup
+          # playwright
```
Comment (TomAFrench) [inline flake.nix:348]: commented code

**Review by TomAFrench — COMMENTED** [review]
Looks fine, some small nits

**Review by kobyhallx — APPROVED** [review]
💯

File: flake.nix (hunk 3 above)
Comment (kobyhallx) [reply flake.nix]: These allow for easier local dev env setup.

**Comment by Savio-Sou** [comment]
This could probably use docs on how to make use of it?
